chainer/functions/loss/ctc.py

```diff
@@ -0,0 +1,247 @@
+from chainer import cuda
+from chainer import function
+from chainer import utils
+import numpy
+
+
+class ConnectionistTemporalClassification(function.Function):
+
+    '''To make it usable for real-world cases, this class has two policies below.
+
+    1. This class computes forward and backward variables in the log domain.
+    2. This class applies the softmax function to inputs. The Backward
+    values of CTC loss is often overflows. This is avoided by computing
+    backward values before the activation function is applied.
+    '''
+
+    def __init__(self, blank_symbol):
+        self.blank_symbol = blank_symbol
+        self.zero_padding = -10000000000.0
+
+    def check_type_forward(self, in_types):
+        utils.type_check.expect(in_types.size() > 1)
+        l_type = in_types[0]
+        utils.type_check.expect(l_type.dtype == numpy.int32)
+
+        x_basetype = in_types[1]
+
+        for i in range(2, len(in_types)):
+            x_type = in_types[1]
+            utils.type_check.expect(
+                x_type.dtype == numpy.float32,
+                x_type.shape == x_basetype.shape,
+            )
+
+    def log_matrix(self, x):
+        xp = cuda.get_array_module(x)
```

Comment:
`cuda.get_array_module` call has overhead.
It is better to make `xp` as an argument of the function.